Clear stale CNP status nodes if updates have been disabled

pippolo84 opened the pull request:
When the option `disable-cnp-status-updates` is set to true, no policy enforcement update is tracked in `CiliumNetworkPolicies`.
However, if the option was previously set to false, the field `status.nodes` still contains the last status of each node when the
feature was turned off.
Currently, the GC in the cilium operator removes status entries only if the relative node has been turned off.

Given that this stale updates information may hinder scalability for large clusters, we extend the GC to clean up all the entries in that field if `disable-cnp-status-updates` is set to true.

Currently, all the errors coming from the GC operations are logged at `debug` level, since the GC is operating as best effort. Besides, being invoked periodically, it could flood the log in case of errors.
We may consider to change that if we want to enhance visibility of errors such the missing permission in ClusterRole.

cc @aanm @gandro 

Fixes #20231

pippolo84 committed:
operator: preallocate cnp list backing array

The number of returned CiliumClusterwideNetworkPolicies is known in
advance, so the preallocation of the backing array will avoid
reallocations after the append.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: fix typos in CNP node status gc

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

aanm commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
```diff
@@ -109,6 +109,7 @@ rules:
   - ciliumclusterwidenetworkpolicies/status
   verbs:
   # Update the auto-generated CNPs and CCNPs status.
+  - patch
```
This commit needs to be backported to all branches because the commit that introduced the regression was also made available to all branches.

aanm commented on operator/cilium_node.go:
```diff
@@ -370,6 +370,21 @@ func runCNPNodeStatusGC(name string, clusterwide bool, nodeStore cache.Store) {
+						// hinder scalability.
+						if option.Config.DisableCNPStatusUpdates {
+							wg.Add(1)
+							cnpCpy := cnp.DeepCopy()
```
Because the `cnpCpy` is never modified inside the `clearStatusNodesInCnp` and because `cnp` is not being retrieved from the k8s-client caches, we don't need to perform a `DeepCopy()` here.

aanm commented on operator/cilium_node.go:
```diff
@@ -370,6 +370,21 @@ func runCNPNodeStatusGC(name string, clusterwide bool, nodeStore cache.Store) {
+						// possible updates written when the option was still enabled.
+						// This is done to avoid the stale updates to accumulate and thus
+						// hinder scalability.
+						if option.Config.DisableCNPStatusUpdates {
```
This entire function, the `runCNPNodeStatusGC`, will only be executed if `operatorOption.Config.CNPNodeStatusGCInterval != 0 `. We advise the users to set `operatorOption.Config.CNPNodeStatusGCInterval` to `0` if they are running `option.Config.DisableCNPStatusUpdates=true`. This effectively means that this code won't be executed and the node status won't be removed.

aanm reviewed (changes requested):
At least the `install/kubernetes: add patch to ciliumnetworkpolicies/status` should be opened in a separate PR so that we can backport that change into all branches.

christarazi commented on operator/cilium_node.go:
```diff
@@ -409,12 +424,28 @@ func runCNPNodeStatusGC(name string, clusterwide bool, nodeStore cache.Store) {
+		},
+	})
+	if err != nil {
+		log.WithError(err).Debug("Unable to json marshal patch object")
```
I think it's a bit clunky to define this here. Rather, having `patchStatusInCNP()` take a `[]k8s.JSONPatch` object seems cleaner to me. That way the marshalling can be done in `patchStatusInCNP()` and if there's an error, then it can be returned.

christarazi commented on operator/cilium_node.go:
```diff
@@ -468,3 +491,19 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+			cnp.Name, types.JSONPatchType, data, meta_v1.PatchOptions{}, "status")
+	} else {
+		_, err = ciliumClient.CiliumNetworkPolicies(namespace).Patch(context.TODO(),
+			cnp.Name, types.JSONPatchType, data, meta_v1.PatchOptions{}, "status")
```
Can we plumb the controller's `context.Context` here instead of `TODO()`?

aanm commented on operator/cilium_node.go:
```diff
@@ -307,6 +307,65 @@ func (c *ciliumNodeUpdateImplementation) Update(origNode, node *cilium_v2.Cilium
+			cnpItemsList = make([]cilium_v2.CiliumNetworkPolicy, 0, len(ccnpList.Items))
+			for _, ccnp := range ccnpList.Items {
+				cnpItemsList = append(cnpItemsList, cilium_v2.CiliumNetworkPolicy{
+					Status: ccnp.Status,
```
This operation won't be necessary for `ClearCnpStatusNodes` because we will remove the status. We will unnecessarily allocate more memory than the one we need just to basically store the CNP name for the `ClearCnpStatusNodes`. Sorry I also wanted to reuse `forEachCNP` but I guess it will need to be specific for the `ClearCnpStatusNodes` and for the `runCNPNodeStatusGC` use cases, unless we perhaps execute the `cb` here.

pippolo84 replied:
Then I guess the best option is to completely separate the code paths and accept a little bit of code duplication.

aanm commented on operator/cilium_node.go:
```diff
@@ -468,3 +489,39 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+			{
+				OP:   "remove",
+				Path: "/status/nodes",
+			},
```
Just to be sure about this, will the operation fail if the `/status/nodes` does not exist in the object or will be a no-op and no error is returned from kubernetes?

pippolo84 replied:
It fails with the following reported error message: `the server rejected our request due to an error in our request`

aanm replied:
Ok, we should then ignore the error or check if the `/status/nodes` ir present with a `op: test` like [here](https://github.com/cilium/cilium/blob/314ca7baeff4f568ffc0bad95124e40665b1f88c/pkg/k8s/cnp.go#L348-L352)

pippolo84 replied:
Reading the [JSON Patch RFC](https://datatracker.ietf.org/doc/html/rfc6902/#section-4.6) it seems that it is not possible to test just for the existence of a field regardless of its value.
Anyway, we can easily ignore the specific error type returned from the client.

aanm commented on operator/cilium_node.go:
```diff
@@ -468,3 +489,39 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+		return err
+	}
+	if err := forEachCNP(ctx, false, cnpsAtOnce, clearStatusNodes); err != nil {
+		return err
```
This is a best-effort operator, whatever errors we get we shouldn't do an early return. Also, there are certain errors that we can ignore, like `k8s.io/apimachinery/pkg/api/errors.IsNotFound` for example.

pippolo84 committed:
operator: use GC controller context while patching CNPs

Use the context from the GC controller to execute the update queries.
Doing so, possible pending queries will be cancelled as soon as the
controller context is cancelled.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: clear CNP status nodes if updates disabled

When the option `disable-cnp-status-updates` is set to true, no policy
enforcement update is tracked in CiliumNetworkPolicies.
However, if the option was previously set to false, the field
status.nodes still contains the last status of each node when the
feature was turned off.
Currently, the GC in the cilium operator removes status entries only if
the relative node has been turned off.

Given that these stale updates may hinder scalability for large
clusters, we clean up all those entries at startup if
`disable-cnp-status-updates` is set to true.

Fixes #20231

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

aanm commented on operator/cilium_node.go:
```diff
@@ -468,3 +467,82 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+			continueID = ccnpList.Continue
+
+			for _, cnp := range ccnpList.Items {
+				_, err = ciliumK8sClient.CiliumV2().CiliumClusterwideNetworkPolicies().Patch(ctx,
```
It is not possible to check with the JSONPatch like you have previously mentioned but we should only perform an API call to kube-apiserver if we know that `len(cnp.status.nodes) != 0`. Doing this for all CNPs can put pressure on the api-server unnecessarily.

aanm commented on operator/cilium_node.go:
```diff
@@ -468,3 +467,82 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+
+			for _, cnp := range cnpList.Items {
+				namespace := utils.ExtractNamespace(&cnp.ObjectMeta)
+				_, err = ciliumK8sClient.CiliumV2().CiliumNetworkPolicies(namespace).Patch(ctx,
```
same here

christarazi reviewed (approved):
LGTM pending Andre's comments

gandro commented on operator/cilium_node.go:
```diff
@@ -467,3 +467,90 @@ func updateCNP(ctx context.Context, ciliumClient v2.CiliumV2Interface, cnp *cili
+	continueID := ""
+	for {
+		if clusterwide {
+			ccnpList, err := ciliumK8sClient.CiliumV2().CiliumNetworkPolicies(core_v1.NamespaceAll).List(
```
I believe this should be `CiliumClusterwideNetworkPolicies`

pippolo84 replied:
Sorry, I messed up the code with a wrong copy-paste. Nice catch!

aanm commented on operator/cilium_node.go:
```diff
@@ -468,3 +467,90 @@ func updateCNP(ciliumClient v2.CiliumV2Interface, cnp *cilium_v2.CiliumNetworkPo
+		if continueID == "" {
+			return
+		}
+	}
```
Only one minor suggestion and is to add a Info log message here informing the user with how many CNPs were cleaned up. Something like `"Garbage collected node status of Cilium Network Policies" found=%d, gc=%d`

gandro reviewed (approved):
Thanks!

pippolo84 committed:
operator: add a flag to skip CNP status cleaning at startup

When the option `disable-cnp-status-updates` is set to true, the
operator, at startup, will garbage collect all stale status nodes
updates in CNPs and CCNPs.

This new option `skip-cnp-status-startup-cleaning` may be used to skip
this clean up so to speed up the operator startup.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: rate limit CNP nodes status clean up

When the option `disable-cnp-status-updates` is set to true, the
operator, at startup, will garbage collect all stale status nodes
updates in CNPs and CCNPs.

To avoid an excessive requests rate to the API server, the clean up is
rate limited. The requests rate per second and the maximum allowed burst
of requests is controlled, respectively, by the two new options
`cnp-status-cleanup-qps` and `cnp-status-cleanup-burst`.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro commented on operator/cilium_node.go:
```diff
@@ -31,6 +31,7 @@ import (
 	nodeStore "github.com/cilium/cilium/pkg/node/store"
 	nodeTypes "github.com/cilium/cilium/pkg/node/types"
 	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/rate"
```
I _think_ we might want to use `golang.org/x/time/rate` here:

See https://github.com/cilium/cilium/pull/12451#discussion_r451601953 for difference between the two packages. The short summary is that `github.com/cilium/cilium/pkg/rate` releases short bursts of updates (e.g. in your configuration, a single burst of 2500 updates (at once) every minute), while `golang.org/x/time/rate` would issue a constant trickle of updates, e.g. 40 updates each second.

To formulate it another way, `cilium/pkg/rate` was designed to fill the target server's queue, to allow it to collect the queued up requests in a batch and process them in one long-running operation. Meanwhile `x/time/rate` is designed to not fill the queue of the target server in the first place. Given that clearing the CNP status is not a long-running operation, and not batched either, I think `x/time/rate` is probably the better choice here.

pippolo84 committed:
chore: fix typo in enableCNPWatcher comment

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
